src/qililab/automatic_calibration/calibration_node.py

```diff
@@ -0,0 +1,682 @@
+
+        Returns:
+            float: The last execution timestamp if exists, None otherwise.
+        """
+        # get all elements in that folder starting with self.node_id+“_” and get the last modified one file name, convert string into datetime
+        last_modified_file_name = self._find_last_executed_calibration()
+        return (
+            os.path.getmtime(f"{self.nb_folder}/{last_modified_file_name}")
+            if last_modified_file_name is not None
+            else None
+        )
+
+    def get_last_calibrated_output_parameters(self) -> dict | None:
+        """Gets the last output of the previous calibration execution of the same notebook if there exist any.
+
+        Returns:
+            dict: The last calibration execution output if exists, None otherwise.
+        """
+        last_modified_file_name = self._find_last_executed_calibration()
+        return (
+            self._parse_output_from_execution_file(last_modified_file_name)
+            if last_modified_file_name is not None
+            else None
+        )
+
+    def _find_last_executed_calibration(self) -> str | None:
+        """Returns the filename of the last calibration execution with the same node_id.
+
+        Returns:
+            str: Name of the file.
+        """
+        # Filtering by node_id and calibrations
+        entries = os.scandir(self.nb_folder)
+        file_names = []
+        for val in entries:
+            if val.is_file():
+                same_node_fnames = val.name.split(f"{self.node_id}_")
+                if len(same_node_fnames) == 2 and len(val.name.split("_calibrated")) == 2:
+                    file_names.append(val.name)
+
+        # Get the last created file, most recent one
+        last_modified_file_time, last_modified_file_name = -1.0, ""
+        for fname in file_names:
+            ftime = os.path.getctime(f"{self.nb_folder}/{fname}")
+            if ftime > last_modified_file_time:
+                last_modified_file_time, last_modified_file_name = ftime, fname
+
+        return last_modified_file_name if last_modified_file_time != -1.0 else None
+
+    def _parse_output_from_execution_file(self, file_name) -> dict | None:
+        """Parses a expected notebook file and gets the output from the execution converted as python dictionary.
+
+        Args:
+            file_name (str): filename of the notebook file.
+
+        Returns:
+            dict: outputs of the executed notebook file.
+
+        Raises:
+            IncorrectCalibrationOutput: In case no outputs, incorrect outputs or multiple outputs where found. Incorrect outputs are those that do not contain `check_parameters` or is empty.
+        """
+        # Parsing file
+        outputs_lines: list[str] = []
+        with open(f"{self.nb_folder}/{file_name}") as file:  # pylint: disable=unspecified-encoding
+            lines = file.readlines()
+            outputs_lines.extend(line for line in lines if line.find(logger_output_start) != -1)
```

Comment:
I think this should be in a `try-catch` block, since we are reading from a file. Also `open(f"{self.nb_folder}/{file_name}")` should be `open(os.path.join(self.nb_folder, file_name), encoding='utf-8')` since it is safer for cross-platform compatibility.

Reply:
But this is for reading, what would happen if the encoding we are trying to read is not utf-8?

Reply:
This is a dicussion, Isaac and me had kind of open still...

Would appreciate some help, but yeah I guess adding the encoding doesn't make what I said before worse..

Reply:
We agreed, we will do as you say! Thanks!

Reply:
Commited!